### PR TITLE
[DC-775] Working min and max pulled from actual dataset

### PR DESCRIPTION
**Pull request by pshapiro4broad**
### Jira Ticket: https://broadworkbench.atlassian.net/browse/dc-775

This is a continuation of work started in https://github.com/DataBiosphere/terra-ui/pull/4493. A new branch and PR was created due to problems with the branch history in the original PR.

## Summary of changes:
This adds a call to the API to get the actual min and max from the dataset listed. This does not add any caching, although the performance seems okay for now with the synthetic data.

Tests to come.

### What
- This adds loading the actual dataset min/max data rather than our stubs
- It also adds loading mechanics for program data criteria.
- Until the criteria is loaded, the user is disallowed from saving the Cohort. This is to avoid trying to store a not-yet-loaded criteria into a cohort.

### Why
- This is important for us so that we are actually starting to load real data

### CODE REVIEW

File: src/dataset-builder/CohortEditor.ts [hunk 1]
```diff
@@ -150,28 +152,43 @@ export const CriteriaView = ({ criteria, deleteCriteria, updateCriteria }: Crite
...
+  return localCriteriaCount;
 };
-
-const createDefaultRangeCriteria = (rangeOption: ProgramDataRangeOption): ProgramDataRangeCriteria => {
-  return {
-    kind: 'range',
-    rangeOption,
-    name: rangeOption.name,
-    id: criteriaCount++,
-    count: 100,
```
Comment (rjohanek) [inline src/dataset-builder/CohortEditor.ts:171]: I remember count is an optional field and I see you don't include it your generated options below. I assume this is because it isn't used or needed anywhere. What was the purpose of the count field originally?

File: src/dataset-builder/CohortEditor.ts [hunk 2]
```diff
@@ -183,19 +200,25 @@ interface DomainOptionWithKind extends DomainOption {
...
+  index: number,
+  option: CriteriaOption
+): Promise<AnyCriteria | undefined> {
   switch (option.kind) {
     case 'list':
-      return createDefaultListCriteria(option);
     case 'range':
-      return createDefaultRangeCriteria(option);
+      return createDefaultProgramDataCriteria(datasetId, index, option);
     case 'domain':
```
Comment (rjohanek) [inline src/dataset-builder/CohortEditor.ts]: it's good to see the union of the range and list criteria and option types being used but I have a question about the domain kind. It seems it isn't fully supported; is this to come later? Or will domain criteria just be handled separately?

File: src/dataset-builder/CohortEditor.ts [hunk 3]
```diff
@@ -206,75 +229,105 @@ type AddCriteriaSelectorProps = {
...
+              value: programDataOption,
+              label: programDataOption.name,
+            };
+          }, snapshotBuilderSettings.programDataOptions),
+        },
+      ],
+      'aria-label': addCriteriaText,
+      placeholder: addCriteriaText,
+      value: null,
+      onChange: async (x) => {
```
Comment (rjohanek) [inline src/dataset-builder/CohortEditor.ts]: is there a better name for x?

File: src/dataset-builder/CohortEditor.ts [hunk 4]
```diff
@@ -206,75 +229,105 @@ type AddCriteriaSelectorProps = {
...
+            if (loadedCriteria !== undefined) {
+              updateCohort((cohort) => {
+                const loading = _.findIndex(
+                  (criteria) => criteria.index === criteriaIndex,
+                  cohort.criteriaGroups[index].criteria
+                );
+                if (loading >= 0) {
+                  return _.set(`criteriaGroups.${index}.criteria.${loading}`, loadedCriteria, cohort);
+                }
+                return cohort;
```
Comment (rjohanek) [inline src/dataset-builder/CohortEditor.ts:277]: So, starting from line 281, the cohort is updated to be loading, then the criteria is loaded, then the cohort is updated again with the loaded criteria.  But I'm a little confused about working of the second update cohort. Specifically, I don't understand the findIndex method.

File: src/dataset-builder/DomainCriteriaSelector.test.ts [hunk 5]
```diff
@@ -47,10 +47,11 @@ describe('DomainCriteriaSelector', () => {
     await user.click(screen.getByLabelText('add'));
     await user.click(screen.getByText('Add to group'));
     // Assert
+    const expectedCriteria = toCriteria(domainOption)(concept);
+    // Update expected index; need to subtract one so indexes match.
+    expectedCriteria.index -= 1;
```
Comment (rjohanek) [inline src/dataset-builder/DomainCriteriaSelector.test.ts]: I see your comment, but why is this necessary?

File: src/libs/ajax/DataRepo.ts [hunk 6]
```diff
@@ -17,26 +17,11 @@ export type SnapshotBuilderDomainOption = {
   root: SnapshotBuilderConcept;
 };
 
-export interface ProgramDataOption {
+export interface SnapshotBuilderProgramDataOption {
   kind: 'range' | 'list';
-  id: number;
```
Comment (rjohanek) [inline src/libs/ajax/DataRepo.ts:27]: Why remove id from options?

File: src/libs/ajax/DatasetBuilder.test.ts [hunk 7]
```diff
@@ -62,15 +60,14 @@ const rangeOption: ProgramDataRangeOption = {
...
-  id: 3,
   name: 'rangeOption',
 };
 
 const rangeCriteria: ProgramDataRangeCriteria = {
   kind: 'range',
   name: 'rangeCriteria',
   rangeOption,
-  id: 4,
+  index: 4,
```
Comment (rjohanek) [inline src/libs/ajax/DatasetBuilder.test.ts:70]: What does the index on a criteria represent? Is this to locate it after loading and to match with the new other type included in `LoadingAnyCriteria`?

**Review by rjohanek — APPROVED** [review]
Looks good, real years are loaded unlike on dev, and the save cohort button is blocked while the range data loads!

File: src/libs/ajax/DataRepo.ts [hunk 8]
```diff
@@ -100,10 +85,50 @@ export interface Snapshot {
...
+interface ColumnStatisticsTextValue {
+  value: string;
+  count: number;
+}
+
 export interface DataRepoContract {
   dataset: (datasetId: string) => {
     details: (include?: DatasetInclude[]) => Promise<DatasetModel>;
     roles: () => Promise<string[]>;
+    lookupDatasetColumnStatisticsById: (
```
Comment (snf2ye) [inline src/libs/ajax/DataRepo.ts]: It would be good to go ahead and switch to the post version of this endpoint, queryDatasetColumnStatisticsById, since we've marked the lookup GET method as deprecated.

File: src/dataset-builder/CohortEditor.ts [hunk 9]
```diff
@@ -183,19 +200,25 @@ interface DomainOptionWithKind extends DomainOption {
...
+  datasetId: string,
+  index: number,
+  option: CriteriaOption
+): Promise<AnyCriteria | undefined> {
   switch (option.kind) {
     case 'list':
-      return createDefaultListCriteria(option);
     case 'range':
-      return createDefaultRangeCriteria(option);
+      return createDefaultProgramDataCriteria(datasetId, index, option);
```
Comment (snf2ye) [inline src/dataset-builder/CohortEditor.ts]: Tagging a bit onto Rae's question below -- It looks like you're switching between these options both here and in createDefaultProgramDataCriteria. Could they be combined?

File: src/dataset-builder/CohortEditor.ts [hunk 10]
```diff
@@ -206,75 +229,105 @@ type AddCriteriaSelectorProps = {
...
+          } else {
+            const criteriaIndex = getNextCriteriaIndex();
+            updateCohort(
+              _.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, {
+                loading: true,
+                index: criteriaIndex,
+              })
+            );
+            const loadedCriteria = await criteriaFromOption(datasetId, criteriaIndex, x.value);
+
```
Comment (snf2ye) [inline src/dataset-builder/CohortEditor.ts]: I'm struggling a bit with reviewing this because it departs significantly from the react style I'm used to. Like the _.set() call above seems like a good candidate for react state and the rest a good candidate for a use of react hooks (like useEffect). But, I think that would be a re-working of this file that's far out of scope for this PR.

File: src/dataset-builder/CohortEditor.ts (hunk 1 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts:171]: Count here is the rollup / participant count for criteria/domain option/cohort/etc. We don't show it for program data criteria. It doesn't really make sense for range criteria; we'd need to know the count for every possible value in the range. You could imagine it being used for list critera, where the UI shows the participant count for different list items, but we don't plan to do that for the MVP release.

File: src/dataset-builder/CohortEditor.ts (hunk 2 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: It's not handled here because this is only used to create a criteria from program data options. The code to create domain critera from an option (aka concept) is the function `toCriteria()` in `DomainCriteriaSelector.ts`. I'll look at trying to make the types reflect this.

File: src/dataset-builder/CohortEditor.ts (hunk 3 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: Yes, it's the selected option type, I can rename this.

File: src/dataset-builder/CohortEditor.ts [hunk 11]
```diff
@@ -150,28 +152,43 @@ export const CriteriaView = ({ criteria, deleteCriteria, updateCriteria }: Crite
...
-    id: Number(_.uniqueId('')),
-    count: 100,
-    values: [listOption.values[0]],
-  };
+const criteriaCount = atom(1);
+export const getNextCriteriaIndex = () => {
+  const localCriteriaCount = criteriaCount.get() + 1;
+  criteriaCount.set(localCriteriaCount);
+  return localCriteriaCount;
 };
```
Comment (snf2ye) [inline src/dataset-builder/CohortEditor.ts]: Ideally, we would track the criteriaIndex in the react state in a location that's shared between both the CohortEditor and DomainCriteriaSelector, which appears to be the DatasetBuilderView. But, I don't see a straightforward path to do this.

File: src/dataset-builder/CohortEditor.ts (hunk 4 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts:277]: The code creates a new unique index for each criteria (using `getNextCriteriaIndex()`) and then appends a placeholder with the index into the criteria group (above on line 283). Once the criteria has loaded, it uses `findIndex()` to look up the actual index in the criteria group by matching on the generated criteria index. This handles the case where the group index may have changed between where the placeholder was inserted due to other insertions or deletions that have occurred while the loading is taking place.

Maybe calling the generated critiera id an `index` is confusing things here, if so I can rename it to something else.

File: src/dataset-builder/DomainCriteriaSelector.test.ts (hunk 5 above)
Comment (pshapiro4broad) [reply src/dataset-builder/DomainCriteriaSelector.test.ts]: Every time a new criteria is created it's assigned a unique index. So each call to `toCriteria()` will result in a new unique index (which is the previous index plus one). The code subtracts one to "back up" the counter so the expected data matches the actual data.

A better approach would be to mock the index creation method to control its results; I can look at doing that for this test.

File: src/libs/ajax/DataRepo.ts (hunk 6 above)
Comment (pshapiro4broad) [reply src/libs/ajax/DataRepo.ts:27]: The original plan was that all options would have an ID, which in the OMOP world corresponds to the concept ID for the property. As it turns out, the program data options are (currently) all values that are stored directly on Person. So there is no concept ID for them; the unique property for them is their name, which the backend can use to map to the database table's column name. So ID isn't needed here.

File: src/libs/ajax/DatasetBuilder.test.ts (hunk 7 above)
Comment (pshapiro4broad) [reply src/libs/ajax/DatasetBuilder.test.ts:70]: Yes, its only purpose is to support async loading of data in the UI. So it's a UI internal value that has no other meaning or use.

File: src/dataset-builder/CohortEditor.ts (hunk 9 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: Interesting, I think there was a need for these to be split before but it looks like they can be combined now. I'll give it a try!

**Review by snf2ye — APPROVED** [review]

File: src/libs/ajax/DataRepo.ts (hunk 8 above)
Comment (pshapiro4broad) [reply src/libs/ajax/DataRepo.ts]: Done, and updated the code to use the new operation id too

File: src/dataset-builder/CohortEditor.ts (hunk 9 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: I was able to merge it into a single method

File: src/dataset-builder/CohortEditor.ts (hunk 10 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: It's not at all apparent by looking at this code, but state is being used here in that `updateCohort()` is implemented by calling a react API (the result of a `useState()` call).

File: src/dataset-builder/CohortEditor.ts (hunk 11 above)
Comment (pshapiro4broad) [reply src/dataset-builder/CohortEditor.ts]: That's a good point; right now this is all still a single page app so instead of using an atom store I could move this to state that's shared by both components. Moving into `DatasetBuilder` could work, let me try doing that.

File: src/dataset-builder/DomainCriteriaSelector.test.ts (hunk 5 above)
Comment (pshapiro4broad) [reply src/dataset-builder/DomainCriteriaSelector.test.ts]: I was able to get rid of the `- 1` by stubbing the criteria index creation function in the test.
